q2_diversity/_ordination.py

```diff
@@ -28,3 +31,67 @@ def pcoa(distance_matrix: skbio.DistanceMatrix,
 def pcoa_biplot(pcoa: skbio.OrdinationResults,
                 features: pd.DataFrame) -> skbio.OrdinationResults:
     return skbio.stats.ordination.pcoa_biplot(pcoa, features)
+
+
+def tsne(distance_matrix: skbio.DistanceMatrix,
+         number_of_dimensions: int = 3,
+         perplexity: float = 30.0,
+         early_exaggeration: float = 12.0) -> skbio.OrdinationResults:
+
+    data = distance_matrix.data
+    ids = distance_matrix.ids
+
+    if number_of_dimensions == 2:
+        number_of_dimensions = 3
+        tsne = TSNE(2, perplexity=perplexity,
+                    early_exaggeration=early_exaggeration).fit_transform(data)
+        add_zeros = np.zeros((tsne.shape[0], 1), dtype=np.int64)
+        tsneData = np.append(tsne, add_zeros, axis=1)
+
+    else:
+        tsneData = TSNE(number_of_dimensions,
+                        perplexity=perplexity,
+                        early_exaggeration=early_exaggeration
+                        ).fit_transform(data)
+
+    axis_labels = ["TSNE%d" % i for i in range(1, number_of_dimensions + 1)]
+    eigenvalues = [0 for i in axis_labels]
+
+    return skbio.OrdinationResults(
+        short_method_name="T-SNE",
+        long_method_name="t-distributed stochastic neighbor embedding",
+        eigvals=pd.Series(eigenvalues, index=axis_labels),
+        proportion_explained=pd.Series(None, index=axis_labels),
+        samples=pd.DataFrame(tsneData, index=ids, columns=axis_labels),
+    )
+
+
+def umap(distance_matrix: skbio.DistanceMatrix,
+         number_of_dimensions: int = 3,
+         n_neighbors: int = 15,
+         min_dist: float = 0.1) -> skbio.OrdinationResults:
+
+    data = distance_matrix.data
+    ids = distance_matrix.ids
+
+    if number_of_dimensions == 2:
+        number_of_dimensions = 3
+        reducer = up.UMAP(n_components=2).fit_transform(data)
+        add_zeros = np.zeros((reducer.shape[0], 1), dtype=np.int64)
+        umap_data = np.append(reducer, add_zeros, axis=1)
+    else:
+        reducer = up.UMAP(n_components=number_of_dimensions,
```

Comment:
reads like a code duplication. Should we wrap this filling of the third dimension into a private function and call from here and within tsne?